packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -112,11 +112,50 @@ export const getLetterEditingOptions = <TDate>(
   }
 };
 
-export const cleanLeadingZeros = <TDate>(
-  utils: MuiPickersAdapter<TDate>,
-  valueStr: string,
-  size: number,
-) => {
+export const getLocalizedDigits = <TDate>(utils: MuiPickersAdapter<TDate>) => {
```

Comment:
The main logic is in those 3 functions.

I build a table with the digits 0 to 9 in the format of the current locale.
And I create 2 functions that allow to convert in both directions (from 0-9 to the localized and from the localized to 0-9).
Then I convert to 0-9 anytime I want to do `Number` or `parseInt` and I convert to localized anytime I want to store the data for rendering.

Reply:
The solution looks simple and elegant. 👍

Reply:
I think I'll add some JSDoc on the methods impacted to say if the value should be in one format or the other to avoid mis-usages